nova_vm/src/ecmascript/types/language/bigint.rs

```diff
@@ -232,35 +232,144 @@ impl BigInt {
+                    Ok(BigInt::SmallBigInt(SmallBigInt(result)))
+                } else {
+                    Ok(agent.heap.create(BigIntHeapData {
+                        data: result.into(),
+                    }))
+                }
+            }
+            (BigInt::SmallBigInt(x), BigInt::BigInt(y))
+            | (BigInt::BigInt(y), BigInt::SmallBigInt(x)) => {
+                let x = x.into_i64();
+                let y = &agent[y];
+                Ok(agent.heap.create(BigIntHeapData { data: x + &y.data }))
+            }
+            (BigInt::BigInt(x), BigInt::BigInt(y)) => {
+                let (x, y) = (&agent[x], &agent[y]);
+                Ok(agent.heap.create(BigIntHeapData {
+                    data: &x.data + &y.data,
+                }))
+            }
+        }
+    }
+
+    /// ### [6.1.6.2.8 BigInt::subtract ( x, y )](https://tc39.es/ecma262/#sec-numeric-types-bigint-subtract)
+    pub(crate) fn subtract(agent: &mut Agent, x: BigInt, y: BigInt) -> JsResult<BigInt> {
+        match (x, y) {
+            (BigInt::SmallBigInt(x), BigInt::SmallBigInt(y)) => {
+                let (x, y) = (x.into_i64() as i128, y.into_i64() as i128);
+                let result = x - y;
+
+                if let Ok(result) = SmallInteger::try_from(result) {
+                    Ok(BigInt::SmallBigInt(SmallBigInt(result)))
+                } else {
+                    Ok(agent.heap.create(BigIntHeapData {
+                        data: result.into(),
+                    }))
+                }
+            }
+            (BigInt::SmallBigInt(x), BigInt::BigInt(y)) => {
+                let x = x.into_i64();
+                let y = &agent[y];
+                Ok(agent.heap.create(BigIntHeapData { data: x - &y.data }))
+            }
+            (BigInt::BigInt(x), BigInt::SmallBigInt(y)) => {
+                let x = &agent[x];
+                let y = y.into_i64();
+                Ok(agent.heap.create(BigIntHeapData { data: &x.data - y }))
+            }
+            (BigInt::BigInt(x), BigInt::BigInt(y)) => {
+                let (x, y) = (&agent[x], &agent[y]);
+                Ok(agent.heap.create(BigIntHeapData {
+                    data: &x.data - &y.data,
+                }))
             }
         }
     }
 
+    /// ### [6.1.6.2.5 BigInt::divide ( x, y )](https://tc39.es/ecma262/#sec-numeric-types-bigint-divide)
+    pub(crate) fn divide(agent: &mut Agent, x: BigInt, y: BigInt) -> JsResult<BigInt> {
+        match (x, y) {
+            (BigInt::SmallBigInt(x), BigInt::SmallBigInt(y)) => {
+                if y == SmallBigInt::zero() {
+                    return Err(agent.throw_exception_with_static_message(
+                        ExceptionType::RangeError,
+                        "Division by zero",
+                    ));
+                }
+                let (x, y) = (x.into_i64() as i128, y.into_i64() as i128);
+                let result = x / y;
```

Comment:
**needs (nonblocking):** A division of two integers will never be smaller in absolute value than the largest of them, so there's not only no chance of overflowing into `i128`, but the result should always be a `SmallBigInt`.